h2o-algos/src/main/java/hex/gam/GAM.java

```diff
@@ -0,0 +1,251 @@
+        Frame newTrain = rebalance(adaptTrain(), false, _result+".temporary.train"); // add and store gam cols without centering
+        dinfo = new DataInfo(newTrain.clone(), _valid, 1, _parms._use_all_factor_levels || _parms._lambda_search, DataInfo.TransformType.NONE, DataInfo.TransformType.NONE,
+                _parms.missingValuesHandling() == GLMParameters.MissingValuesHandling.Skip,
+                _parms.missingValuesHandling() == GLMParameters.MissingValuesHandling.MeanImputation || _parms.missingValuesHandling() == GLMParameters.MissingValuesHandling.PlugValues,
+                _parms.makeImputer(),
+                false, hasWeightCol(), hasOffsetCol(), hasFoldCol(), null);
+        DKV.put(dinfo._key, dinfo);
+        model = new GAMModel(dest(), _parms, new GAMModel.GAMModelOutput(GAM.this, dinfo._adaptedFrame, dinfo));
+        model.delete_and_lock(_job);
+        newTFrame = _centerGAM?new Frame(buildGamFrameCenter(_parms._gam_X.length, _gamFrameKeysCenter, _parms.train(), 
+                _parms._response_column, _parms.train().numCols(), _parms.train().names())):new Frame(dinfo._adaptedFrame);  // get frames with correct predictors and spline functions
+        DKV.put(newTFrame); // This one will cause deleted vectors if add to Scope.track
+        if (_parms._saveGamCols) {  // save gam column keys
+          gamColsFrame = saveGAMFrames(newTrain);
+          DKV.put(gamColsFrame);  // do not Scope.track this, will cause null frame
+          model._output._gamTransformedTrain = gamColsFrame._key;
+          if (_centerGAM) {
+            gamColsFrameCenter = saveGAMFrames(newTFrame);
+            DKV.put(gamColsFrameCenter);
+            model._output._gamTransformedTrainCenter = gamColsFrameCenter._key;
+            model._output._gamGamXCenter = newTFrame._key;
+          }
+        }
+        
+        _job.update(1, "calling GLM to build GAM model...");
+        GLMModel glmModel = buildGLMModel(_parms, newTFrame); // obtained GLM model
+        Scope.track_generic(glmModel);
+        _job.update(0, "Building out GAM model...");
+        fillOutGAMModel(glmModel, model, dinfo); // build up GAM model 
+        
+        // call adaptTeatForTrain() to massage frame before scoring.
+        // call model.makeModelMetrics
+        // create model summary by calling createModelSummaryTable or something like that
+        model.update(_job);
+        // build GAM Model Metrics
+      } finally {
+        List<Key<Vec>> keep = new ArrayList<>();
+        if (model != null) {
+          if (_parms._saveGamCols) {
+            addFrameKeys2Keep(newTFrame._key, keep);
+            addFrameKeys2Keep(model._output._gamTransformedTrain, keep);
+            if (_centerGAM)
+              addFrameKeys2Keep(model._output._gamTransformedTrainCenter, keep);
+          }
+          model.unlock(_job);
+        }
+        if (dinfo!=null)
+          dinfo.remove();
+        Scope.untrack(keep);  // leave the vectors alone.
```

Comment:
The `Scope.untrack` can be moved closer to the scope of the `keep` variable.

Reply:
I could propose writing the whole block like that:

```patch
Index: h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GamUtils.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GamUtils.java	(revision cfbe780da3f15b91ff29fd39314198c537837571)
+++ h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GamUtils.java	(date 1583247280868)
@@ -319,9 +319,11 @@
     return gamFrameCenter;
   }
   
-  public static void addFrameKeys2Keep(Key<Frame> keyNames, List<Key<Vec>> keep) {
-      Frame loadingFrm = DKV.getGet(keyNames);
+  public static void addFrameKeys2Keep(List<Key<Vec>> keep, Key<Frame> ... keyNames) {
+    for (Key<Frame> keyName : keyNames) {
+      Frame loadingFrm = DKV.getGet(keyName);
       if (loadingFrm != null) for (Vec vec : loadingFrm.vecs()) keep.add(vec._key);
+    }
   }
   
   public static Frame saveGAMFrames(Frame fr) {
Index: h2o-algos/src/main/java/hex/gam/GAM.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- h2o-algos/src/main/java/hex/gam/GAM.java	(revision cfbe780da3f15b91ff29fd39314198c537837571)
+++ h2o-algos/src/main/java/hex/gam/GAM.java	(date 1583247315596)
@@ -206,16 +206,16 @@
         List<Key<Vec>> keep = new ArrayList<>();
         if (model != null) {
           if (_parms._saveGamCols) {
-            addFrameKeys2Keep(newTFrame._key, keep);
-            addFrameKeys2Keep(model._output._gamTransformedTrain, keep);
+            addFrameKeys2Keep(keep, model._output._gamTransformedTrain);
             if (_centerGAM)
-              addFrameKeys2Keep(model._output._gamTransformedTrainCenter, keep);
+              addFrameKeys2Keep(keep, model._output._gamTransformedTrainCenter);
           }
           model.unlock(_job);
-        }
-        if (dinfo!=null)
-          dinfo.remove();
-        Scope.untrack(keep);  // leave the vectors alone.
+          Scope.untrack(keep);  // leave the vectors alone.
+        }
+        if (dinfo != null) {
+          dinfo.remove();
+        }
       }
     }
     
```

But it's really optional. Maybe you'll like it.

Reply:
Pavel:  I incorporated everything here.  Very elegant!